File: conformance/code/individual/src/main/java/com/digit/individual/observability/BusinessMetrics.java
package com.digit.individual.observability;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import org.springframework.stereotype.Component;

/**
 * Individual-specific business metrics. Mirrors Go pkg/observability/business_metrics.go.
 *
 * <p>Backed by the Micrometer {@link MeterRegistry} provided by Spring Boot Actuator (the same
 * registry the official tracer's ObservabilityMetrics uses), so these counters surface alongside the
 * tracer's http_server_requests / db_operations metrics on /actuator/metrics + /actuator/prometheus.
 */
@Component
public class BusinessMetrics {

    private final MeterRegistry registry;

    public BusinessMetrics(MeterRegistry registry) {
        this.registry = registry;
    }

    public void recordIndividualCreated(String tenantId, int count) {
        Counter.builder("individuals_created_total")
                .description("Total number of individuals created")
                .tag("tenantId", tenantId)
                .tag("operation", "create")
                .register(registry)
                .increment(count);
    }

    public void recordIndividualSearched(String tenantId, int resultCount) {
        Counter.builder("individuals_searched_total")
                .description("Total number of individual searches performed")
                .tag("tenantId", tenantId)
                .tag("operation", "search")
                .register(registry)
                .increment();
    }

    public void recordIndividualUpdated(String tenantId, int count) {
        Counter.builder("individuals_updated_total")
                .description("Total number of individuals updated")
                .tag("tenantId", tenantId)
                .tag("operation", "update")
                .register(registry)
                .increment(count);
    }

    public void recordIndividualDeleted(String tenantId, int count) {
        Counter.builder("individuals_deleted_total")
                .description("Total number of individuals deleted (soft)")
                .tag("tenantId", tenantId)
                .tag("operation", "delete")
                .register(registry)
                .increment(count);
    }

    public void recordConfigUpserted(String tenantId, boolean created) {
        Counter.builder("configs_upserted_total")
                .description("Total number of tenant configs upserted")
                .tag("tenantId", tenantId)
                .tag("operation", created ? "create" : "update")
                .register(registry)
                .increment();
    }
}

File: conformance/code/individual-go/pkg/observability/business_metrics.go
package observability

import (
	"context"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

var (
	individualsCreated  metric.Int64Counter
	individualsSearched metric.Int64Counter
	individualsUpdated  metric.Int64Counter
	individualsDeleted  metric.Int64Counter
	configsUpserted     metric.Int64Counter
)

// InitializeBusinessMetrics creates individual-specific business metrics
func InitializeBusinessMetrics() error {
	meter := otel.Meter("individual-service")

	var err error

	individualsCreated, err = meter.Int64Counter(
		"individuals_created_total",
		metric.WithDescription("Total number of individuals created"),
	)
	if err != nil {
		return err
	}

	individualsSearched, err = meter.Int64Counter(
		"individuals_searched_total",
		metric.WithDescription("Total number of individual searches performed"),
	)
	if err != nil {
		return err
	}

	individualsUpdated, err = meter.Int64Counter(
		"individuals_updated_total",
		metric.WithDescription("Total number of individuals updated"),
	)
	if err != nil {
		return err
	}

	individualsDeleted, err = meter.Int64Counter(
		"individuals_deleted_total",
		metric.WithDescription("Total number of individuals deleted (soft)"),
	)
	if err != nil {
		return err
	}

	configsUpserted, err = meter.Int64Counter(
		"configs_upserted_total",
		metric.WithDescription("Total number of tenant configs upserted"),
	)
	if err != nil {
		return err
	}

	return nil
}

func RecordIndividualCreated(ctx context.Context, tenantID string, count int) {
	attrs := []attribute.KeyValue{
		attribute.String("tenantId", tenantID),
		attribute.String("operation", "create"),
	}
	individualsCreated.Add(ctx, int64(count), metric.WithAttributes(attrs...))
}

func RecordIndividualSearched(ctx context.Context, tenantID string, resultCount int) {
	attrs := []attribute.KeyValue{
		attribute.String("tenantId", tenantID),
		attribute.String("operation", "search"),
		attribute.Int("result_count", resultCount),
	}
	individualsSearched.Add(ctx, 1, metric.WithAttributes(attrs...))
}

func RecordIndividualUpdated(ctx context.Context, tenantID string, count int) {
	attrs := []attribute.KeyValue{
		attribute.String("tenantId", tenantID),
		attribute.String("operation", "update"),
	}
	individualsUpdated.Add(ctx, int64(count), metric.WithAttributes(attrs...))
}

func RecordIndividualDeleted(ctx context.Context, tenantID string, count int) {
	attrs := []attribute.KeyValue{
		attribute.String("tenantId", tenantID),
		attribute.String("operation", "delete"),
	}
	individualsDeleted.Add(ctx, int64(count), metric.WithAttributes(attrs...))
}

func RecordConfigUpserted(ctx context.Context, tenantID string, created bool) {
	op := "update"
	if created {
		op = "create"
	}
	attrs := []attribute.KeyValue{
		attribute.String("tenantId", tenantID),
		attribute.String("operation", op),
	}
	configsUpserted.Add(ctx, 1, metric.WithAttributes(attrs...))
}
